Reading capabilities from disk

thenor opened the pull request.

tobsan commented on agent/CMakeLists.txt:
```diff
@@ -25,6 +25,8 @@ add_executable(softwarecontainer-agent
     src/softwarecontaineragentadaptor.cpp
     ${GENERATED_ADAPTOR_FILENAME}
+    src/capability/configstore.cpp
+    src/capability/capability.cpp
```
This class still does nothing.

thenor replied:
I know. Kept partly to start discussion on if I should use it. If no reason pops up for using it, I will remove it.

thenor replied:
Removed in next upload.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+ReturnCode ConfigStore::readCapsFromFile(const std::string &filePath)
+{
+    if (!is_json(filePath)) {
+        log_debug() << "File is not a json file: \n" << filePath;
```
log_ macros will append newlines automatically, and you probably don't want newline before the filePath anyway. This goes for all uses of these log macros.

thenor replied:
Will fix. Update in next upload.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+
+GatewayConfiguration ConfigStore::readConfigs(const std::string capID)
+{
+    return m_configStore[capID];
```
Perhaps a warning about non-existing cap in case the cap is not in the store?

thenor replied:
Sounds very reasonable.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+    size_t i;
+    json_t *capability;
+
+    json_array_foreach(capabilities,i,capability) {
```
Would be nice with spaces here

thenor replied:
Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+        }
+
+        std::string capName;
+        if (!JSONParser::read(capability,"name",capName)) {
```
Spaces here also.

thenor replied:
Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+            return ReturnCode::FAILURE;
+        }
+
+        return parseGateways(capName,gateways);
```
Space please

thenor replied:
Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+
+ReturnCode ConfigStore::parseGateways(std::string capName, json_t *gateways)
+{
+    size_t j;
```
Just personal preference, but I would prefer if this was called i.

thenor replied:
Forgot this when I made it into a help method.

thenor replied:
Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+        }
+
+        std::string gwID;
+        if (!JSONParser::read(gateway,"id",gwID)) {
```
Spaces please

thenor replied:
Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+            return ReturnCode::FAILURE;
+        }
+
+        json_t *confs = json_object_get(gateway,"config");
```
Spaces please.

thenor replied:
Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+     * same ID, then append the configs from this file to the
+     * existing Gateway in the m_configstore
+     */
+    for (auto it=gwConfs.begin(); it!=gwConfs.end(); ++it) {
```
Using C++11 style for loop would be nicer here:

```
for (auto const &it : gwConfs) {

}
```

thenor replied:
Looks magical. Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+    } else {
+        struct dirent *hFile;
+        while ((hFile = readdir(dirFile)) != nullptr) {
+            // TODO (thenor): we should check for '.' and '..'
```
Perhaps this TODO should be removed (since it checks for ".json" later anyway)

thenor replied:
This was in part a reminder about looking in to how to handle ".." or ".". How do you think they should be handled?
I think other checks will make this fail or not fail anyway. Was just thinking if I should handle this as "/" where I don't allow it.

tobsan replied:
Currently there is a function that checks if the file name ends with ".json". The special files "." and ".." does not end with ".json", so this shouldn't be a problem.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -31,6 +59,7 @@ SoftwareContainerAgent::~SoftwareContainerAgent()
 
 bool SoftwareContainerAgent::triggerPreload()
 {
+    //log_debug() << "triggerPreload " << m_preloadCount - m_preloadedContainers.size();
```
Remove this or have it uncommented.

thenor replied:
Done. Mistake during pull from upstream.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
+        std::string capId = capabilities[i];
+        GatewayConfiguration newConfigs = m_configStore.readConfigs(capId);
+
+        setGatewayConfigs(containerID,newConfigs);
```
Space here, please.

thenor replied:
Done.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
+        GatewayConfiguration newConfigs = m_configStore.readConfigs(capId);
+
+        setGatewayConfigs(containerID,newConfigs);
+    }
```
This loop is actually a nice solution so that one doesn't have to merge the configs manually.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
+{
+    GatewayConfiguration parsedConfigs;
+
+    for (auto it=configs.begin(); it!=configs.end(); ++it) {
```
You could use a c++11 style loop here as well.

thenor replied:
Yes. Done.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
+bool SoftwareContainerAgent::setCapabilities(const uint32_t &containerID,
+                                             const std::vector<std::string> &capabilities)
 {
+    for (unsigned int i = 0; i != capabilities.size(); i++) {
```
A range based loop would be nice here as well.

thenor replied:
Done.

tobsan commented on agent/src/softwarecontaineragent.h:
```diff
@@ -235,5 +248,6 @@ class SoftwareContainerAgent : protected softwarecontainer::JSONParser
     SignalConnectionsHandler m_connections;
     bool m_shutdownContainers = true;
     std::vector<ContainerID> m_containerIdPool;
+    ConfigStore m_configStore = ConfigStore("");
```
This is not very nice, create the m_configStore in the constructor instead.

thenor replied:
I agree, but couldn't get it to compile then. I will grab you to see if you can see a solution.

thenor replied:
Done.

tobsan commented on agent/src/softwarecontaineragentadaptor.cpp:
```diff
@@ -9,7 +9,14 @@ softwarecontainer::SoftwareContainerAgentAdaptor::SoftwareContainerAgentAdaptor(
+    const std::string &workingDirectory,
+    const std::string &outputFile,
+    const std::map<std::string,
+    std::string> &env)
```
Don't put one argument on two lines, please.

thenor replied:
Oops. Done.

tobsan commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -82,37 +82,45 @@ TEST_F(SoftwareContainerAgentTest, CreateContainerWithConf) {
 
-    ASSERT_TRUE(retval == true);
-    ASSERT_TRUE(workspace->m_enableWriteBuffer == false);
+    // TODO (thenor): Most likely gives a false positive. Find a way to fix this.
```
Yes, please.

thenor replied:
Done. Hopefully....

tobsan commented:
Also please rebase against master

breakreturn commented on agent/src/capability/configstore.h:
```diff
@@ -0,0 +1,79 @@
+
+        std::map<std::string, GatewayConfiguration> m_configStore;
+    };
+}
```
We usually don't indent the namespace, unindent the class level.

thenor replied:
Done.

breakreturn commented on agent/src/capability/configstore.h:
```diff
@@ -0,0 +1,79 @@
+        /**
+         * @brief Returns all Gateways for a certain Capability.
+         */
+        GatewayConfiguration readConfigs(const std::string capID);
```
I think the above comments should be rewritten to not use the word "gateway" like it is currently. A "Gateway" is a specific thing, a "gateway configuration" is different specific thing.

thenor replied:
Oh yes. Of course. Done.

breakreturn commented on agent/src/main.cpp:
```diff
@@ -10,12 +29,13 @@ void usage(const char *argv0)
     printf("--user <uid>        : Default user id to be used when starting processes in the container, defaults to 0\n");
     printf("--shutdown <bool>   : If false, containers will not be shutdown on exit. Useful for debugging. Defaults to true\n");
     printf("--timeout <seconds> : Timeout in seconds to wait for containers to shutdown, defaults to 2\n");
+    printf("--configPath <path> : Path to a file or directory where Service Manifests exist, defaults to \"\"");
```
We might very well add a "SoftwareContainer configuration" concept later, so I think we should use an option name that is more specific for this. Like, service manifest path or something like that.

thenor replied:
Done.

breakreturn commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -1,14 +1,34 @@
+    const std::string &configPath)
+    : m_mainLoopContext(mainLoopContext)
+    , m_preloadCount(preloadCount)
+    , m_shutdownContainers(shutdownContainers)
```
This indentation makes it difficult to see the difference between the parameters and the initialization of members.

thenor replied:
Agreed. Done.

breakreturn commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
 
-bool SoftwareContainerAgent::setCapabilities(const uint32_t &containerID, const std::vector<std::string> &capabilities)
+void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID,
+                                               const std::map<std::string, std::string> &configs)
```
Some line break inconsistency here.

thenor replied:
Actually not since it's a map of strings.

DunderRoffe commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+
+    if (fileroot == nullptr) {
+        log_error() << "Could not parse Service Manifest: \n" <<
+            filePath << ":" << error.line <<"\n" << error.text;
```
The `<<` operator should be before `filePath` not last on the line above.

thenor replied:
Done.

breakreturn commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
+        std::string gwID = it->first;
+        std::string configStr = it->second;
+
+        json_t *jConfigs = json_loads(configStr.c_str(),0,&error);
```
Put spaces between args

thenor replied:
Done.

breakreturn commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
 }
 
-bool SoftwareContainerAgent::setCapabilities(const uint32_t &containerID, const std::vector<std::string> &capabilities)
+void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID,
```
Perhaps we should return a bool from this as we do with setCapabilities?

thenor replied:
Yes, I had a thought about this also. Does @tobsan agree?

tobsan replied:
Yes, that makes sense.

thenor replied:
I have done this, but feel like it may require even more changes to the softwarecontainer class, so let's see if this is sufficient. Changes in next upload.

DunderRoffe commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+
+    }
+
+    GatewayConfiguration oldGwConfs = m_configStore[capName];
```
Maybe return `ReturnCode::FAILURE` if there is no conf for `capName` in `m_configStore`

thenor replied:
This method is also used when parsing a file from the constructor, when the m_configstore is empty, so that would mean that it fails as soon as we try to add a new capability.

tobsan replied:
Isn't the point here to **not fail** even if there are no old configs. The [] operator on std::map constructs an object of the value type if there is none available. I think it can be kept as is.

If there is a nice way to make it explicit if there are any old configs or not without making it too complicated, that could also be nice of course.

thenor replied:
I will add this as an improvement note, but unless you feel I need to change something now, I will not change this.

DunderRoffe commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,237 @@
+
+    DIR *dirFile = opendir(filePath.c_str());
+    if (dirFile == nullptr) {
+        log_error() << "Could not open directory: " << filePath << "\n";
```
Return empty `std::vector< ... >` here. Then we do not need the else clause and can lower the indentation.

thenor replied:
Cool. Done.

breakreturn commented on agent/unit-test/configstore_unittest.cpp:
```diff
@@ -0,0 +1,119 @@
+    std::string sm_path       = "CS_unittest_ServiceManifest.json";
+    std::string short_sm_path = "CS_unittest_short_ServiceManifest.json";
+    std::string evil_sm       = "CS_unittest_parseError.json";
+    std::string inputDir      = "/home/vagrant/";
```
We need to have some way of not having the paths absolute to the vagrant machine.

thenor replied:
I thought about this also, but I wasn't sure how. If I had tested this from e.g. the Agent it would have felt more intuitive to test the default container paths also. More tests are needed.

breakreturn commented on agent/unit-test/configstore_unittest.cpp:
```diff
@@ -0,0 +1,119 @@
+
+TEST_F(ConfigStoreTest, readEvilConfig) {
+    ASSERT_TRUE(isError(cs->readCapsFromFile(homeDir+evil_sm)));
+}
```
In general: add comments of what is tested, i.e. why would some calls be false etc. Also add spaces around operators like '+'

thenor replied:
Done.

DunderRoffe commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -261,8 +303,38 @@ void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, cons
+        std::string configStr = it->second;
+
+        json_t *jConfigs = json_loads(configStr.c_str(),0,&error);
+        if (!jConfigs) {
```
You alternate between using `!ptr` and `ptr == nullptr` please choose one style.

thenor replied:
Is there any case where !ptr could be a bad syntax? In my opinion it's cleaner since it's shorter. Any opinion to the contrary, @DunderRoffe ?

DunderRoffe replied:
No there are no functional difference between the two as far as I am aware.

thenor replied:
I will use "!ptr". Looks nicer.

DunderRoffe commented on agent/unit-test/configstore_unittest.cpp:
```diff
@@ -0,0 +1,119 @@
+public:
+    LOG_DECLARE_CLASS_CONTEXT("TEST", "Tester");
+    ConfigStoreTest() { }
+    ConfigStore *cs = new ConfigStore("");
```
This does not seem to be deleted anywhere.

thenor replied:
Done.

breakreturn commented on common/jsonparser.cpp:
```diff
@@ -24,7 +24,8 @@ namespace softwarecontainer {
 {
     json_t *value = json_object_get(element, key);
-    if (!value) {
+    if (value == nullptr) {
```
If this is a better check, should the other similar checks below be changed too?

thenor replied:
DunderRoffe also commented on this. I will change the "ptr == nullptr" to "!ptr".

DunderRoffe reviewed (changes requested):
Overall good. Found a few things, a bit picky but still worth considering.

breakreturn commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -252,7 +293,8 @@ std::string SoftwareContainerAgent::bindMountFolderInContainer(const uint32_t co
 
-void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID, const std::map<std::string, std::string> &configs)
+void SoftwareContainerAgent::setGatewayConfigs(const uint32_t &containerID,
+                                               const GatewayConfiguration &configs)
```
As far as I can see one method "setGatewayConfigs" is used by the user and the other overloaded version is not intended to be used by the user, rather by the other method. In this case one of them should be private, and I suggest we change the name and not only overload. Unless we plan for the user to be able to use both? I wouldn't be a fan of that either unless there's a good reason though.

thenor replied:
Yes, I wasn't sure about this, but I agree. What does @tobsan say about this change?

tobsan replied:
I agree, make the internal one private (the other one is used by the DBus API)

thenor replied:
Done.

breakreturn commented on agent/src/capability/configstore.h:
```diff
@@ -0,0 +1,79 @@
+    private:
+
+        ReturnCode parseCapabilities(json_t *capabilities);
+        ReturnCode parseGateways(std::string capName, json_t *gateways);
```
We parse "gateway configs" and not "gateways", or am I missing the point?

thenor replied:
This method actually parses the gateways in the json file, so in this case it is actually a "gateways" and not gateway configurations.

thenor replied:
I will change the name.

breakreturn commented on agent/src/capability/configstore.h:
```diff
@@ -0,0 +1,79 @@
+         * @return ReturnCode::FAILURE if parsing of files is unsuccessful
+         *
+         */
+        ReturnCode readCapsFromDir(const std::string &dirPath);
```
What's the intention for the user of this class? In this design, should anything apart from getting the actual gateway configs by cap ID be public?

thenor replied:
The reason I made reading from file/dir public was that I thought perhaps users would want to start up the container and agent first, then read in the configuration. Perhaps not a valid use case?

breakreturn commented on agent/src/capability/configstore.h:
```diff
@@ -0,0 +1,79 @@
+        /**
+         * @brief Returns all Gateways for a certain Capability.
+         */
+        GatewayConfiguration readConfigs(const std::string capID);
```
We don't really do any "reading" at this stage, perhaps this method name should be changed so it doesn't accidentally imply implementation details.

thenor replied:
Done.

breakreturn commented on common/softwarecontainer-common.h:
```diff
@@ -40,7 +42,7 @@ static constexpr ContainerID INVALID_CONTAINER_ID = -1;
 static constexpr int INVALID_FD = -1;
 
-typedef std::map<std::string, std::string> GatewayConfiguration;
+typedef std::map<std::string, json_t *> GatewayConfiguration;
```
I still think this should be a class so we can hide raw types from external dependencies more easily.

thenor replied:
I will add that to my todo list. I think it's a good idea.

tobsan commented on agent/unit-test/CS_unittest_ServiceManifest.json:
```diff
@@ -0,0 +1,80 @@
+{
```
Could we place this in some test-data directory instead?

thenor replied:
Done.

thenor commented on agent/src/capability/capability.cpp:
```diff
@@ -0,0 +1,32 @@
+
```
Removed.

thenor commented on agent/src/capability/capability.h:
```diff
@@ -0,0 +1,44 @@
+
```
Removed.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,242 @@
+        return ReturnCode::SUCCESS;
+    }
+
+    for (unsigned int i = 0; i != files.size(); i++) {
```
You can use a range-loop here.

thenor replied:
Done.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -277,17 +315,54 @@ std::string SoftwareContainerAgent::bindMountFolderInContainer(const uint32_t co
+    for (std::string capId : capabilities) {
+        GatewayConfiguration newConfigs = m_configStore.getGatewayConfigs(capId);
+
+        if (!updateGatewayConfigs(containerID, newConfigs)) {
```
An error log would be nice here.

thenor replied:
Perhaps a "return" here may have been wrong.
Do we want to fail if we can't set the configuration for one capability?

tobsan replied:
We are doing it differently in different parts of the code actually. I think it should be a failure if we can't apply our configs.

thenor replied:
Done.

tobsan commented on agent/unit-test/configstore_unittest.cpp:
```diff
@@ -0,0 +1,172 @@
+public:
+    LOG_DECLARE_CLASS_CONTEXT("TEST", "Tester");
+    ConfigStoreTest() { }
+    ConfigStore *cs = new ConfigStore();
```
Couldn't this be created on the stack instead, to avoid having to do delete in the teardown?

thenor replied:
Is this what @DunderRoffe did for the agent unittest? I can look at that code and do that here also.

DunderRoffe replied:
What @tobsan means is that it should not be a pointer.

thenor replied:
Done.

tobsan commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -36,18 +35,20 @@ class SoftwareContainerAgentTest: public ::testing::Test
     int m_shutdownTimeout = 2;
     std::shared_ptr<Workspace> workspace;
+    std::string m_configPath =
+        "/home/vagrant/softwarecontainer/agent/unit-test/test-data/CS_unittest_ServiceManifest.json";
```
Make this a relative path.

thenor replied:
I thought that would be the bad way to go. Why is a relative path better? So that I can move the test case and the test-data directory? I was more concerned that the tests could be started from "not here" and the relative path would point incorrectly.

tobsan replied:
It will be equally bad if someone doesn't build softwarecontainer in vagrant, or in some other path in the VM. We can't assume everyone will build in a VM. You should be able to point to the tests relative to the test binary. It is reasonable to assume that test-data directory is available in the same place as the test binary.

thenor replied:
Alright. Will change it.

thenor commented on agent/src/capability/configstore.h:
```diff
@@ -0,0 +1,111 @@
+    ConfigStore(const std::string &filePath);
+    ~ConfigStore();
+
+    /**
```
After discussions with @breakreturn I will make these two read methods private instead. They are not relevant to have public for any current use case.

DunderRoffe commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,250 @@
+    if (inputPath.empty()) {
+        //File path is empty, this is ok
+        return;
+    } else if (isDirectory(inputPath)) {
```
In the cases were we return in an if statement, I think it is nice to not use an else clause. In cases such as this, consider using another if statement. This makes it separated and more clear that the function (or constructor in this case) terminates.

thenor replied:
Done.

tobsan commented on agent/src/capability/configstore.cpp:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    std::vector<std::string> files;
+    files = fileList(dirPath);
```
This can be just one line instead.

thenor replied:
Done.

DunderRoffe reviewed (commented):
Looks good overall.
There are some cases where you need to change to use  "== nullptr" checks.

DunderRoffe commented on libsoftwarecontainer/src/softwarecontainer.cpp:
```diff
@@ -163,8 +157,9 @@ void SoftwareContainer::setGatewayConfigs(const GatewayConfiguration &configs)
-            std::string config = configs.at(gatewayId);
-            gateway->setConfig(config);
+            json_t *config = configs.at(gatewayId);
+            char *configStr = json_dumps(config,0);
```
All that is returned by json_dumps needs to be freed

thenor replied:
Done.

thenor commented on agent/unit-test/configstore_unittest.cpp:
```diff
@@ -0,0 +1,172 @@
+    std::string capNameD = "dummyCapD";
+
+    /* The Service Manifest (absolute) file paths */
+    std::string homeDir       = "/home/vagrant/softwarecontainer/agent/unit-test/test-data/";
```
I need to make this into a relative path as discussed.

thenor replied:
Done.

Commit:
Reading capabilities from disk

Parse capability from Service Manifest files and save them internally.
Can read from file or from directory.

Adapt SoftwareContainer to use ConfigStore
    * Changes how Gateway configurations are represented,
      so the rest of SC is adapted to use this representation.
    * Add instantiation of ConfigStore object in the Agent
Unit tests for ConfigStore
    * Test that it is possible to read configurations from Service Manifests.
    * Verify that configurations are merged if several files contain
      configuration for the same Capability (based on capability name).

thenor commented:
One new unit test case fails in Jenkins, but didn't fail in my vagrant, so I will try to reproduce this locally.
